File: internal/forms/federal/schedule_se.go
package federal

import (
	"math"

	"taxpilot/internal/forms"
)

func init() { forms.RegisterForm(ScheduleSE) }

// Self-employment tax constants for 2025
const (
	ssTaxRate2025          = 0.124  // Social Security portion (12.4%)
	medicareTaxRate2025    = 0.029  // Medicare portion (2.9%)
	seTaxRate2025          = 0.9235 // 92.35% of net SE earnings are taxable
	ssWageBase2025         = 176100 // Social Security wage base for 2025
	additionalMedicareRate = 0.009  // Additional Medicare Tax rate (0.9%)
	additionalMedicareBase = 200000 // Threshold for Additional Medicare Tax (single)
)

// ScheduleSE returns the FormDef for Schedule SE — Self-Employment Tax.
// This computes the self-employment tax (Social Security + Medicare) on
// net earnings from self-employment (Schedule C line 31).
func ScheduleSE() *forms.FormDef {
	return &forms.FormDef{
		ID:           forms.FormScheduleSE,
		Name:         "Schedule SE — Self-Employment Tax",
		Jurisdiction: forms.Federal,
		TaxYears:      []int{2025},
		QuestionGroup: "business",
		QuestionOrder: 5,
		Fields: []forms.FieldDef{
			// Line 2: Net earnings from self-employment (from Schedule C)
			forms.RefField("2", "Net earnings from self-employment", forms.SchedCLine31),
			// Line 3: 92.35% of line 2 (if > $400)
			{
				Line:      "3",
				Type:      forms.Computed,
				Label:     "Self-employment earnings subject to tax",
				DependsOn: []string{forms.SchedSELine2},
				Compute: func(dv forms.DepValues) float64 {
					net := dv.Get(forms.SchedSELine2)
					if net < 400 {
						return 0 // SE tax only applies if >= $400
					}
					return net * seTaxRate2025
				},
			},
			// Line 4: Social Security tax portion
			// 12.4% on earnings up to the wage base ($176,100 for 2025),
			// reduced by W-2 Social Security wages.
			{
				Line:      "4",
				Type:      forms.Computed,
				Label:     "Social Security tax",
				DependsOn: []string{forms.SchedSELine3, forms.W2WildcardSSWages},
				Compute: func(dv forms.DepValues) float64 {
					seEarnings := dv.Get(forms.SchedSELine3)
					if seEarnings <= 0 {
						return 0
					}
					w2SSWages := dv.SumAll(forms.W2WildcardSSWages)
					remainingBase := math.Max(0, ssWageBase2025-w2SSWages)
					taxableForSS := math.Min(seEarnings, remainingBase)
					return taxableForSS * ssTaxRate2025
				},
			},
			// Line 5: Medicare tax portion (2.9% on all SE earnings, no cap)
			{
				Line:      "5",
				Type:      forms.Computed,
				Label:     "Medicare tax",
				DependsOn: []string{forms.SchedSELine3},
				Compute: func(dv forms.DepValues) float64 {
					seEarnings := dv.Get(forms.SchedSELine3)
					if seEarnings <= 0 {
						return 0
					}
					return seEarnings * medicareTaxRate2025
				},
			},
			// Line 6: Total self-employment tax
			forms.SumField("6", "Self-employment tax", forms.SchedSELine4, forms.SchedSELine5),
			// Line 7: Deductible part (50% of SE tax — goes to Schedule 1 line 16)
			{
				Line:      "7",
				Type:      forms.Computed,
				Label:     "Deductible part of self-employment tax",
				DependsOn: []string{forms.SchedSELine6},
				Compute: func(dv forms.DepValues) float64 {
					return dv.Get(forms.SchedSELine6) * 0.5
				},
			},
		},
	}
}

File: api/src/forms/federal/schedule_se.rs
use crate::domain::builders::*;
use crate::domain::field::*;
use crate::domain::form::*;

// Self-employment tax constants for 2025
const SS_TAX_RATE: f64 = 0.124;
const MEDICARE_TAX_RATE: f64 = 0.029;
const SE_TAX_RATE: f64 = 0.9235;
const SS_WAGE_BASE_2025: f64 = 176_100.0;

pub fn schedule_se() -> FormDef {
    FormDef {
        id: FORM_SCHEDULE_SE.to_string(),
        name: "Schedule SE — Self-Employment Tax".to_string(),
        jurisdiction: Jurisdiction::Federal,
        tax_years: vec![2024, 2025, 2026],
        question_group: "business".to_string(),
        question_order: 5,
        fields: vec![
            // Line 2
            ref_field("2", "Net earnings from self-employment", SCHED_C_LINE_31),
            // Line 3: 92.35% of line 2 (if > $400)
            {
                let deps = vec![SCHED_SE_LINE_2.to_string()];
                FieldDef::new_computed("3", "Self-employment earnings subject to tax", deps, Box::new(|dv: &DepValues| {
                    let net = dv.get(SCHED_SE_LINE_2);
                    if net < 400.0 { 0.0 } else { net * SE_TAX_RATE }
                }))
            },
            // Line 4: Social Security tax portion
            {
                let deps = vec![
                    SCHED_SE_LINE_3.to_string(),
                    W2_WILDCARD_SS_WAGES.to_string(),
                ];
                FieldDef::new_computed("4", "Social Security tax", deps, Box::new(|dv: &DepValues| {
                    let se_earnings = dv.get(SCHED_SE_LINE_3);
                    if se_earnings <= 0.0 {
                        return 0.0;
                    }
                    let w2_ss_wages = dv.sum_all(W2_WILDCARD_SS_WAGES);
                    let remaining_base = (SS_WAGE_BASE_2025 - w2_ss_wages).max(0.0);
                    let taxable_for_ss = se_earnings.min(remaining_base);
                    taxable_for_ss * SS_TAX_RATE
                }))
            },
            // Line 5: Medicare tax
            {
                let deps = vec![SCHED_SE_LINE_3.to_string()];
                FieldDef::new_computed("5", "Medicare tax", deps, Box::new(|dv: &DepValues| {
                    let se_earnings = dv.get(SCHED_SE_LINE_3);
                    if se_earnings <= 0.0 { 0.0 } else { se_earnings * MEDICARE_TAX_RATE }
                }))
            },
            // Line 6: Total SE tax
            sum_field("6", "Self-employment tax", vec![SCHED_SE_LINE_4, SCHED_SE_LINE_5]),
            // Line 7: Deductible part (50%)
            {
                let deps = vec![SCHED_SE_LINE_6.to_string()];
                FieldDef::new_computed("7", "Deductible part of self-employment tax", deps, Box::new(|dv: &DepValues| {
                    dv.get(SCHED_SE_LINE_6) * 0.5
                }))
            },
        ],
    }
}
